File: csharp/demo/demo10-yield.cs
/*
 * THREAD YIELDING
 */
using System;
using System.Threading;



class Demo10 : IRunnable
{
    public void run()
    {
        DateTime tpStartMeasure = DateTime.Now;

        littleSleep(1);

        var timeElapsed = DateTime.Now.Subtract(tpStartMeasure).TotalMilliseconds;

        Console.WriteLine($"Elapsed time: {timeElapsed} miliseonds");
    }


    private void littleSleep(double miliseconds)
    {
        DateTime tpEnd = DateTime.Now.AddMilliseconds(miliseconds);

        do
        {
            Thread.Yield();
        }
        while (DateTime.Now.CompareTo(tpEnd) < 0);
    }
}

File: cpp-std/demo10-yield.cpp
/*
THREAD YIELDING
*/


#include <iostream>
#include <thread>
#include <chrono>
#include "mytool-time.hpp"
using namespace std;



using chrmicro = std::chrono::microseconds;
using hrclock = mytool::HiResClock;



void littleSleep(int us) {
    auto tpStart = hrclock::now();
    auto tpEnd = tpStart + chrmicro(us);

    do {
        std::this_thread::yield();
    }
    while (hrclock::now() < tpEnd);
}



int main() {
    auto tpStartMeasure = hrclock::now();

    littleSleep(130);

    auto timeElapsed = hrclock::getTimeSpan<chrmicro>(tpStartMeasure);

    cout << "Elapsed time: " << timeElapsed.count() << " microseonds" << endl;
    return 0;
}
